distributed/shuffle/_worker_plugin.py

```diff
@@ -323,44 +320,58 @@ def __init__(
         self.worker_for = worker_for
         self.split_axes = split_axes(old, new)
 
-    async def _receive(self, data: list[tuple[ArrayRechunkShardID, bytes]]) -> None:
+    async def _receive(self, data: list[tuple[NDIndex, bytes]]) -> None:
         self.raise_if_closed()
 
-        buffers = defaultdict(list)
+        filtered = []
         for d in data:
             id, payload = d
             if id in self.received:
                 continue
+            filtered.append(payload)
             self.received.add(id)
             self.total_recvd += sizeof(d)
-
-            buffers[id.chunk_index].append(payload)
-
         del data
-        if not buffers:
+        if not filtered:
             return
         try:
-            await self._write_to_disk(buffers)
+            shards = await self.offload(self._repartition_shards, filtered)
+            del filtered
+            await self._write_to_disk(shards)
         except Exception as e:
             self._exception = e
             raise
 
+    def _repartition_shards(self, data: list[bytes]) -> dict[NDIndex, bytes]:
+        repartitioned: defaultdict[
+            NDIndex, list[tuple[NDIndex, np.ndarray]]
+        ] = defaultdict(list)
+        for buffer in data:
+            for tpl in pickle.loads(buffer):
+                id, shard = tpl
+                repartitioned[id].append(shard)
+        return {k: pickle.dumps(v) for k, v in repartitioned.items()}
```

Comment:
I think this kind of operation should be done by the buffers. This would require a bit of a bigger rewrite but would likely also solve https://github.com/dask/distributed/issues/7990

Reply:
Agreed, that's been on my pile of "wants" for a while, but as you said, it requires a more extensive rewrite of the buffers, so it's out of scope for this PR. The focus here is to reduce the footprint of `ShuffleRun.received` without negatively impacting runtime performance.